Stop setting force_noinline for functions of union arguments

Keno opened the pull request:
The original motivation for setting force_noinline for functions
with union arguments seems to have been bugs in type intersection.
However, the type system has improved quite a bit and at least in
the test suites we don't seem to run into any issues. Thus, we can
stop setting force_noinline. Nevertheless, it still make since
to penalize functions with union arguments. Doing so encourages
union splits to happen at the call site, rather than having multiple
redundant union splits inside the function. However, for simple functions
of intrinsics and builtins it can make a lot of sense to inline
the union split signature.

Keno commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/27057#issuecomment-387933363) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/f2487b44fcd824f49803b30c995293691597ce24/9afded9_vs_e4fa75d/report.md). cc @ararslan

vtjnash commented:
Didn't something like this get merged?

Keno commented:
Did this actually get merged? I don't remember that happening.

vtjnash commented:
oh, apparently not. why do we have that check in there?

vtjnash commented:
bump

vtjnash commented:
bump

vtjnash commented:
bump?

Keno committed:
Stop setting force_noinline for function of union arguments

The original motivation for setting force_noinline for functions
with union arguments seems to have been bugs in type intersection.
However, the type system has improved quite a bit and at least in
the test suites we don't seem to run into any issues. Thus, we can
stop setting force_noinline. Nevertheless, it still make since
to penalize functions with union arguments. Doing so encourages
union splits to happen at the call site, rather than having multiple
redundant union splits inside the function. However, for simple functions
of intrinsics and builtins it can make a lot of sense to inline
the union split signature.
